compiler/rustc_middle/src/ty/layout.rs

```diff
@@ -3182,7 +3183,39 @@ impl<'tcx> LayoutCx<'tcx, TyCtxt<'tcx>> {
                 }
 
                 match arg.layout.abi {
-                    Abi::Aggregate { .. } => {}
+                    Abi::Aggregate { .. } => {
+                        // Pass and return structures up to 2 pointers in size by value,
+                        // matching `ScalarPair`. LLVM will usually pass these in 2 registers
+                        // which is more efficient than by-ref.
+                        let max_by_val_size = Pointer.size(self) * 2;
+                        let size = arg.layout.size;
+
+                        if arg.layout.is_unsized() || size > max_by_val_size {
+                            arg.make_indirect();
+
+                        // If the data is small enough and homogenous we want to aggregates it to
+                        // gain some register and put less pression on the codegen.
+                        } else if let Ok(HomogeneousAggregate::Homogeneous(reg)) =
+                            arg.layout.homogeneous_aggregate(self)
+                        {
+                            match &reg.kind {
+                                // We want to pass small aggregates as immediates, but using
+                                // a LLVM aggregate type for this leads to bad optimizations,
+                                // so we pick an appropriately sized integer type instead.
+                                RegKind::Integer => {
+                                    arg.cast_to(Reg { kind: RegKind::Integer, size })
+                                }
+                                // We don't want to aggregate float as Integer because this will
+                                // hurt (#93490) the codegen and performance so instead use a
+                                // vector (this is similar to what clang is doing in C++).
+                                RegKind::Float => arg.cast_to(Reg { kind: RegKind::Vector, size }),
```

Comment:
This is unsound. Different `-Ctarget-features` of different crates can cause ABI incompatibilities.

Reply:
Given the comment below, that was also what I taught but trying my PR with a code that use different target features leads me to the correct output:
```rust
#![feature(bench_black_box)]
#![allow(non_camel_case_types)]

#[derive(Debug)]
struct f32x4(f32, f32, f32, f32);

#[target_feature(enable = "avx")]
unsafe fn foo() -> f32x4 { std::hint::black_box(f32x4(0., 1., 2., std::hint::black_box(3.))) }

#[target_feature(enable = "sse3")]
unsafe fn bar(arg: f32x4) {
    println!("{:?} == f32x4(0.0, 1.0, 2.0, 3.0)", std::hint::black_box(arg));
}

fn main() { unsafe { bar(std::hint::black_box(foo())); } }

``` 
```text
$ rustc +stage1 -C opt-level=0 a.rs && ./a
f32x4(0.0, 1.0, 2.0, 3.0) == f32x4(0.0, 1.0, 2.0, 3.0)
$ rustc +stage1 -C opt-level=3 a.rs && ./a
f32x4(0.0, 1.0, 2.0, 3.0) == f32x4(0.0, 1.0, 2.0, 3.0)
```

Reply:
That's just a 128bit vector which fits into a single XMM register. And since that is the lower half of the YMM register in AVX code, that probably just happens to work by accident on x86_64.

I guess using an x86 target with code that has the SSE target feature enabled on only foo or bar might fail (careful, IIRC the i686 arch enables SSE2bby default). Or maybe some of the other supported archs can trigger a bug here as well.

Reply:
The i586 targets disable SSE by default I believe. You can also manually disable it.

Reply:
@dotdash Yes and no, above this code there is a check for the max size by value who is ptr-size * 2, so the created vector cannot be (at least on x86_64) upper than 128bits which as you said fits on the lower half of the YMM register. This is also consistent with clang++ who (accordingly to my test) never pass vectors more than ptr-size * 2 by value.

Reply:
@Urgau that's because the SysV x86_64 specifies it like that. But the Rust ABI is (currently) non-arch specific so it has to work for every supported arch. That the 128bit limit makes this work for x86_64 is (a far as I'm concerned) by accident, because that limit is not there to make this work, but to limit the size of aggregates that get passed in registers by any means. It's meant to make a tradeoff between stack and register usage in function calls.

clang++ using the SysV x86_64 ABI will happily pass a vector of more than 128bits by value if you ask for such a vector explicitly (https://godbolt.org/z/ejcfKKW5j). It just doesn't automatically promote any aggregate to vectors because that would violate the ABI.

Reply:
I think this is now resolved by the fact that this PR only does the cast to Vector in x86_64.

Reply:
This is a breaking change. This would break returning a struct of floats using the rust abi if SSE is disabled and soft-float isn't enabled. In that case LLVM will refuse to use the xmm registers, but the abi requires it to use them, so it throws an error.

```rust
pub struct f32x4(f32, f32, f32, f32);

pub extern "C" fn foo() -> f32x4 {
    f32x4(1.0, 2.0, 3.0, 4.0)
}
```

gives

```
error: <source>:5:2: in function _ZN7example3foo17h753a09441bc16561E { double, double } (): SSE register return with SSE disabled
```

when compiled with `-Ctarget-feature=-sse`. https://rust.godbolt.org/z/f4Gehcnx7 This PR changes the rust abi to do the exact same thing. While you could argue that `-sse` should be used in combination with `+soft-float` and that returning a bare float doesn't work either before this PR, it is still technically a breaking change.

Reply:
@bjorn3 I don't really follow, the example you gave use `extern "C"` but my PR only modify the Rust ABI.
And if I test your example with the Rust ABI and my changes it compiles with both `-Ctarget-feature=-sse` and `-Ctarget-feature=-sse,+soft-float`, so I don't think their is any breaking changes.

Reply:
> I don't really follow, the example you gave use extern "C" but my PR only modify the Rust ABI.

Because I expected that current rust would compile `extern "C"` the same as the rust abi with this PR for this example.

> And if I test your example with the Rust ABI and my changes it compiles with both -Ctarget-feature=-sse and -Ctarget-feature=-sse,+soft-float, so I don't think their is any breaking changes.

Huh, what llvm-ir does rust produce for `-Ctarget-feature=-sse` with this PR if you omit `extern "C"`?

Reply:
Oh, it does perform a vector promotion. But IIRC the responsible code doesn't expect this to happen and only provides a `u8` vector type of the proper total size. IIRC it was you that answered a question related to this on Zulip a while back, when I tried to get the SysV ABI to actually use a `<2xf32>` instead of a single `f64` when passing aggregates of `f32` values.

Reply:
Right, looks like it doesn't error. It however does change the abi from returning the value in xmm versus on the stack. This means that this makes disabling SSE support UB where previously it wasn't, which is still a breaking change.

Reply:
Noooo... Well this is bad, really bad. @bjorn3 Any recommendation(s) on how to proceed from now ?

Reply:
Not sure. For vector types we "fix" this by forcing them to always be passed on the stack for the rust abi and hoping llvm's mem2reg pass will eliminate the stack usage after inlining.

Reply:
@bjorn3 I've updated the PR to not modify the Reg representation of homogeneous floats. I think this now completely  resolve your concern about ABI compatibility as the generated IR doesn't use a vector type and is the same as rust <= 1.47.
For information as before it doesn't compile with `-Ctarget-feature=-sse` but does compile with `-C target-feature=-sse,+soft-float`.

Reply:
Are you satisfied with the changes, @bjorn3?